crates/core/src/rpc/surfnet_cheatcodes.rs

```diff
@@ -617,24 +618,89 @@ impl SvmTricksRpc for SurfnetCheatcodesRpc {
                     context: RpcResponseContext::new(0),
                     value: ProfileResult {
                         compute_units: error_cu_result,
+                        state: ProfileState::new(BTreeMap::new(), BTreeMap::new()),
                     },
                 }));
             }
         };
 
         Box::pin(async move {
+            let account_keys_to_profile = transaction.message.static_account_keys().to_vec();
+            // In a future step, we could extend account_keys_to_profile with an optional RPC parameter `accounts_to_profile`
+
+            let mut pre_execution_capture = BTreeMap::new();
+            for key in &account_keys_to_profile {
+                let SvmAccessContext {
+                    inner: account_result,
+                    ..
+                } = svm_locker.get_account_local(key);
+                match account_result {
+                    GetAccountResult::FoundAccount(_, acc, _) => {
+                        pre_execution_capture.insert(*key, Some(acc.data.clone()));
+                    }
+                    GetAccountResult::FoundProgramAccount(
+                        (prog_key, prog_acc),
+                        (_prog_data_key, _prog_data_acc_option),
+                    ) => {
+                        // get_account_local typically returns FoundAccount even for program executables.
+                        log::trace!(
+                            "Capturing program executable account {} data during pre-exec.",
+                            prog_key
+                        );
+                        pre_execution_capture.insert(prog_key, Some(prog_acc.data.clone()));
+                    }
+                    GetAccountResult::None(_) => {
+                        pre_execution_capture.insert(*key, None);
+                    }
+                }
+            }
+
             let SvmAccessContext {
                 slot,
-                inner: estimation_result,
+                inner: estimation_result, // This is ComputeUnitsEstimationResult
                 ..
             } = svm_locker.estimate_compute_units(&transaction);
 
+            // TODO: This currently fetches the original account state again, as
```

Comment:
I want to be sure I understand this TODO correctly - does our `estimate_compute_units` not actually update the pre/post state, so the result of fetching the accounts again won't actually be shown in the profile?

Reply:
@BretasArthur1, I never saw an answer to this question. I know the TODO is gone, but I'm still curious of the answer - can we actually see a difference in the before/after state here?